### PR TITLE
Hwol mlnx dpdk17.11

### PR DESCRIPTION
**Pull request by OphirMunk**
Implement flow aging and statistics

### CODE REVIEW

**Comment by OphirMunk** [comment]
Pull request was updated. @asafpenso @rbaryanai please review

File: lib/netdev-dpdk.c [hunk 1]
```diff
@@ -4069,6 +4068,35 @@ ufid_to_rte_flow_disassociate(const ovs_u128 *ufid) {
               UUID_ARGS((struct uuid *)ufid));
 }
 
+static int
+netdev_dpdk_flow_stats_get(struct netdev *netdev, const ovs_u128 *ufid,
```
Comment (rbaryanai) [inline lib/netdev-dpdk.c]: This function should be implemented in netdev-rte-offloads.c

File: lib/netdev-dpdk.c [hunk 2]
```diff
@@ -4069,6 +4068,35 @@ ufid_to_rte_flow_disassociate(const ovs_u128 *ufid) {
...
+                     struct dpif_flow_stats *stats) {
+
+    int ret = 0;
+    struct rte_flow_error error;
+    struct rte_flow_query_count query;
+    struct netdev_dpdk *dev = netdev_dpdk_cast(netdev);
+
+    struct rte_flow *rte_flow = ufid_to_dpdk_rte_flow(ufid);
```
Comment (rbaryanai) [inline lib/netdev-dpdk.c]: might have many RTE flows, correct only in the default case.

File: lib/netdev-rte-offloads.c [hunk 3]
```diff
@@ -692,6 +692,29 @@ int netdev_vport_flow_del(struct netdev * netdev OVS_UNUSED,
...
+    rte_port = netdev_rte_port_search(port_no, &dpdk_map);
+
+    if (rte_port == NULL) {
+        VLOG_ERR("failed to find dpdk port for port %d",port_no);
+        return NULL;
+    }
+
+    struct ufid_hw_offload *ufid_hwol = ufid_hw_offload_find(ufid, &rte_port->ufid_to_rte);
+    if (ufid_hwol) {
+        return ufid_hwol->rte_flow_data[0].flow;
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.c]: correct only if we have one rte_flow.

File: lib/netdev-rte-offloads.h [hunk 4]
```diff
@@ -53,4 +53,6 @@ int netdev_dpdk_flow_put(struct netdev *netdev, struct match *match,
 int netdev_dpdk_flow_del(struct netdev *netdev, const ovs_u128 *ufid,
                      struct dpif_flow_stats *stats OVS_UNUSED);
 
+struct rte_flow *
+ufid_to_dpdk_rte_flow(const ovs_u128 *ufid);
```
Comment (rbaryanai) [inline lib/netdev-rte-offloads.h]: should not be exposed.

File: lib/netdev-rte-offloads.c [hunk 5]
```diff
@@ -692,6 +692,29 @@ int netdev_vport_flow_del(struct netdev * netdev OVS_UNUSED,
     return 0;
 }
 
+struct rte_flow *
+ufid_to_dpdk_rte_flow(const ovs_u128 *ufid) {
```
Comment (asafpenso) [inline lib/netdev-rte-offloads.c]: probably should be static since it's not defined in header file

File: lib/netdev-dpdk.c [hunk 6]
```diff
@@ -4069,6 +4068,35 @@ ufid_to_rte_flow_disassociate(const ovs_u128 *ufid) {
...
+    if (!rte_flow) {
+        return EINVAL;
+    }
+    memset(&query, 0, sizeof(query));
+    /* reset counters after query */
+    query.reset = 1;
+    ret = rte_flow_query(dev->port_id, rte_flow, RTE_FLOW_ACTION_TYPE_COUNT,
+                         &query, &error);
+    if (ret) {
```
Comment (asafpenso) [inline lib/netdev-dpdk.c]: I suggest doing if (!ret) stats->n_packets etc... 
This way you'll have a single return ret.

File: lib/dpif-netdev.c [hunk 7]
```diff
@@ -3564,6 +3564,38 @@ dpif_netdev_flow_dump_thread_destroy(struct dpif_flow_dump_thread *thread_)
...
+    odp_port_t in_port = netdev_flow->flow.in_port.odp_port;
+
+    ovs_mutex_lock(&pmd->dp->port_mutex);
+    port = dp_netdev_lookup_port(pmd->dp, in_port);
+    ovs_assert(port);
+    if (!port) {
+            ovs_mutex_unlock(&pmd->dp->port_mutex);
+            return -1;
+    }
+    /* get offloaded stats */
```
Comment (asafpenso) [inline lib/dpif-netdev.c]: the code is self explanatory so i think the comment can be omitted

File: lib/netdev-rte-offloads.c (hunk 5 above)
Comment (OphirMunk) [reply lib/netdev-rte-offloads.c]: It is defined in netdev-rte-offloads.h

File: lib/netdev-dpdk.c (hunk 6 above)
Comment (OphirMunk) [reply lib/netdev-dpdk.c]: Thanks but it seems more readable this way and I prefer not to indent a big block of code. This style is also exercised throughout the file so the current code is the correct style.

File: lib/dpif-netdev.c (hunk 7 above)
Comment (OphirMunk) [reply lib/dpif-netdev.c]: Agreed. Fixed.

**Pull request merged**